core/checksums/src/main/java/software/amazon/awssdk/checksums/internal/DigestAlgorithm.java

```diff
@@ -27,41 +31,92 @@ public enum DigestAlgorithm {
     }
 
     public String getAlgorithmName() {
         return algorithmName;
     }
 
     /**
-     * Returns the thread local reference for the {@link MessageDigest} algorithm
+     * Returns a {@link CloseableMessageDigest} to use for this algorithm.
      */
-    public MessageDigest getDigest() {
-        MessageDigest digest = digestReference.get();
-        digest.reset();
-        return digest;
+    public CloseableMessageDigest getDigest() {
+        MessageDigest digest = digestCache.pollFirst();
+        if (digest != null) {
+            digest.reset();
+            return new CloseableMessageDigest(digest);
+        }
+        return new CloseableMessageDigest(newDigest());
+    }
+
+    private MessageDigest newDigest() {
+        try {
+            return MessageDigest.getInstance(algorithmName);
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("Unable to fetch message digest instance for Algorithm "
+                                       + algorithmName + ": " + e.getMessage(), e);
+        }
     }
 
-    private static class DigestThreadLocal extends ThreadLocal<MessageDigest> {
-        private final String algorithmName;
+    public final class CloseableMessageDigest implements SdkAutoCloseable, Cloneable {
+        private final Supplier<MessageDigest> closedDigest = () -> {
+            throw new IllegalStateException("This message digest is closed.");
+        };
+        private Supplier<MessageDigest> digest;
+        private byte[] messageDigest;
+
+        private CloseableMessageDigest(MessageDigest digest) {
+            this.digest = () -> digest;
+        }
+
+        /**
+         * Retrieve the message digest instance.
+         */
+        public MessageDigest messageDigest() {
+            return digest.get();
+        }
 
-        DigestThreadLocal(String algorithmName) {
-            this.algorithmName = algorithmName;
+        /**
+         * Retrieve the message digest bytes. This will close the message digest when invoked. This is because the underlying
+         * message digest is reset on read, and we'd rather fail future interactions with the digest than act on the wrong data.
+         */
+        public byte[] digest() {
+            if (messageDigest != null) {
+                return messageDigest;
+            }
+            messageDigest = messageDigest().digest();
+            return messageDigest;
+        }
+
+        /**
+         * Release this message digest back to the cache. Once released, you must not use the digest anymore.
+         */
+        @Override
+        public void close() {
+            if (digest == closedDigest) {
+                return;
+            }
+
+            // Avoid over-caching after large traffic bursts. The maximum chosen here is arbitrary. It's also not strictly
+            // enforced, since these statements aren't synchronized.
+            if (digestCache.size() <= MAX_CACHED_DIGESTS) {
```

Comment:
size() may be expensive, should we track the size using atomic integer instead?

Reply:
How would you coordinate the atomic size and the concurrent deque?

Reply:
I'll switch to a linked blocking deque. Uses locking, but it's still likely faster than creating new checksums and has a constant-time size() method. We can do benchmarking later to verify this is the fastest method.

Is that reasonable?